frontend/ui/widget/doublespinwidget.lua

```diff
@@ -28,26 +29,30 @@ local DoubleSpinWidget = InputContainer:new{
     width = nil,
     width_factor = nil, -- number between 0 and 1, factor to the smallest of screen width and height
     height = nil,
+    left_text = _("Left"),
     left_min = 1,
     left_max = 20,
     left_value = 1,
     left_default = nil,
-    left_text = _("Left"),
+    left_precision = nil, -- default "%02d" in NumberPickerWidget
+    left_wrap = false,
+    right_text = _("Right"),
     right_min = 1,
     right_max = 20,
     right_value = 1,
     right_default = nil,
-    right_text = _("Right"),
+    right_precision = nil,
+    right_wrap = false,
     cancel_text = _("Close"),
     ok_text = _("Apply"),
     cancel_callback = nil,
     callback = nil,
     close_callback = nil,
     keep_shown_on_apply = false,
-    -- Set this to add default button that restores numbers to their default values
-    default_values = nil,
-    default_text = _("Use defaults"),
-    -- Optional extra button on bottom
+    -- Set this to add upper default button that applies default values with callback(nil, nil)
+    default_values = false,
+    default_text = nil,
+    -- Optional extra button above ok/cancel buttons row
     extra_text = nil,
     extra_callback = nil,
```

Comment:
So, "default" applies immediately ?
If we want a default that does not apply immediately, we have to use extra and extra_callback ?
You asked about that and I answered at https://github.com/koreader/koreader/pull/8488#issuecomment-981097540 - is this becoming the generic behaviour (apply immediately) then ? If yes, I'm not sure it's for the best, vs. just updating the number picker value (apply would then be just an additional tap).

Also, not deep into the spin widgets as you are - but are the behaviours (and default behaviour) of single and double spin widget similar ? (It feels single Spin widget do not apply immediately ?)

Reply:
Before this PR default buttons in SW and DSW worked differently, and there I asked shouldn't the SW behaviour be changed. You recommended no, so - no changes.
In SW the default button just sets the default value to the numberpicker, in DSW default values are applied immediately with `callback(nil, nil)`. (This nil goes from the default settings for hyphenation limits).
There are 3 DSWs with the default button only.

Reply:
Well, playing again with hyphenation limits, I think I wouldn't mind if the "Use language defaults" did not apply immediately, and would just update the values. This feature is probably really rarely used, so I think we can change this behaviour without risk (but we can still make this widget prettier and a little less wide :)

DSW indeed comes from this hyphenation limits, which was initially just a local widget in readerhyphenation.lua.
It's possible that at the time, I did not know the "languages defaults" values, so I had to apply immediately to be able to fetch them (this might have been the reason for `(nil, nil)`). This seem to be no longer the case (and the comment in there about language.json is obsolete).

Unless keeping `(nil, nil)` is just fine as a way to "apply immediately".
Was going to suggest we may have another option `defaults_apply_immediately` in both SW and DSW, so we can get the best feature depending on needs :)
So, if you feel like this `(nil, nil)` could be cleaned up and is not really needed by any other DSW user, you can try making them more similar in look and code usage.

Reply:
Now the "language defaults" are saved as `nil`.
https://github.com/koreader/koreader/blob/1895d61eda2afe8b7f3acfb70ea8c0a6b5ad1244/frontend/apps/reader/modules/readertypography.lua#L386-L388
If the default button behaviour is changed, the "language default" values will be put into the numberpickers and then on `Apply` saved as values, not `nil`s.

Reply:
Oh, then, that might have been the reason for (nil, nil), that we indeed save "nil" !
No matter what the UI behaviour is, I'd rather like the settings to be saved as `nil` (so not be there in the settings), when we decide to use language defaults. If we change the default values for a language in crengine, we want to use the new values, and not be stucked with old values for this language.

Reply:
And if a user changes only one side, we want to save both values (meaning it's not the language defaults), even if the other side stays the same as the language default for this side.

Reply:
So, let the default button in DSW be applied immediately, the default text in it shows the fact.

Reply:
As you feel is right :)

Dunno how consistent this would be with our other widgets, or if it could be a new way to do things - but we could have the "Apply/OK/Set" button grayed out when the picker value is actually the current one, so when "Use defaults" would change the values, or after any manual change, we would get "Apply" change from gray to black to indicate one should tap it to really apply the values.

Reply:
> the "Apply/OK/Set" button grayed out when the picker value is actually the current one

I've made it working in SpinWidget via broadcasting an event `NumberPickerUpdated`.
If it's not an overkill, I can propose (separately).
Maybe another idea how to signal from the NumberPicker to the parent SpinWidget?

Reply:
Well, for this kind of things, broadcastEvent is really overkill (and a bit "not the right way to do it", as some other picker could be in the way catching it).
Usually, we just pass a reference or a callback down from the top level widget (that may want to be notified of something happening in a lower widget) to some/all lower widget, like `parent=` or `dialog=`.
It's already used for refreshes/repaint, so may be it's already there and you can re-use it?
If not (or if it doesn't reference the widget you want), you could just add another one: `widget_with_apply_button`, but a callback sounds probably better: `picker_updated_callback`

Reply:
Thanks, got it.